src/main/java/com/salesforce/dynamodbv2/mt/util/RecordsCache.java

```diff
@@ -0,0 +1,276 @@
+            assert start.compareTo(from) <= 0 && end.compareTo(from) >= 0;
+
+            if (records.isEmpty()) {
+                return records;
+            }
+
+            if (start.equals(from)) {
+                return records;
+            }
+
+            return records.subList(getIndex(from), records.size());
+        }
+
+        private int getIndex(SequenceNumber sequenceNumber) {
+            final List<SequenceNumber> sequenceNumbers = Lists.transform(records, SequenceNumber::fromRecord);
+            int index = Collections.binarySearch(sequenceNumbers, sequenceNumber);
+            if (index < 0) {
+                index = (-index) - 1;
+            }
+            return index;
+        }
+
+        Segment subSegment(Segment previousSegment, Segment nextSegment) {
+            return subSegment(
+                previousSegment == null ? null : previousSegment.getEnd().next(),
+                nextSegment == null ? null : nextSegment.getStart()
+            );
+        }
+
+        Segment subSegment(SequenceNumber from, SequenceNumber to) {
+            assert from == null || to == null || from.compareTo(to) < 0;
+
+            int cf = from == null ? 1 : start.compareTo(from);
+            int cl = to == null ? -1 : end.compareTo(to);
+
+            if (cf >= 0) {
+                // first sequence number of this segment is after from: start with first
+                if (cl <= 0) {
+                    // last sequence number of this segment is before to: end with last
+                    return this;
+                } else {
+                    // last sequence number of this segment is after to: end with to
+                    final List<Record> newRecords = new ArrayList<>(records.subList(0, getIndex(to)));
+                    return new Segment(start, to, newRecords);
+                }
+            } else {
+                // first sequence number of this segment if before from: start with from
+                if (cl <= 0) {
+                    // last sequence number of this segment is before to: end with last
+                    final List<Record> newRecords = new ArrayList<>(records.subList(getIndex(from), records.size()));
+                    return new Segment(from, end, newRecords);
+                } else {
+                    // last sequence number of this segment is after to: end with to
+                    final List<Record> newRecords = new ArrayList<>(records.subList(getIndex(from), getIndex(to)));
+                    return new Segment(from, to, newRecords);
+                }
+            }
+        }
+
+        long getByteSize() {
+            return byteSize;
+        }
+
+        boolean isEmpty() {
+            return start.equals(end);
+        }
+    }
+
+    // config parameter
+    private final long maxRecordsByteSize;
+
+    // cached record segments sorted by sequence number within each shard
+    private final Map<ShardId, NavigableMap<SequenceNumber, Segment>> segments;
```

Comment:
This is another behavior change (non-functional) compared to master: instead of using a single sorted map where the key class includes streamArn, shardId, and SequenceNumber, this impl uses a hashmap on streamArn+shardId and a nested sorted map on sequenceNumber. That should reduce the amount of computation needed to compare and lookup segments.